x-pack/plugins/alerting/server/alerts_client/alerts_client.ts

```diff
@@ -0,0 +1,377 @@
+> {
+  logger: Logger;
+  ruleType: UntypedNormalizedRuleType;
+  elasticsearchClientPromise: Promise<ElasticsearchClient>;
+  legacyAlertsClient: LegacyAlertsClient<
+    LegacyState,
+    LegacyContext,
+    ActionGroupIds,
+    RecoveryActionGroupId
+  >;
+}
+
+// todo would be nice to infer this from AlertSchema
+export interface AlertRuleSchema {
+  [ALERT_RULE_CATEGORY]: string;
+  [ALERT_RULE_CONSUMER]: string;
+  [ALERT_RULE_EXECUTION_UUID]: string;
+  [ALERT_RULE_NAME]: string;
+  [ALERT_RULE_PRODUCER]: string;
+  [ALERT_RULE_TAGS]: string[];
+  [ALERT_RULE_TYPE_ID]: string;
+  [ALERT_RULE_UUID]: string;
+  [ALERT_RULE_PARAMETERS]: unknown;
+  [SPACE_IDS]: string[];
+}
+
+interface InitializeOpts {
+  rule: {
+    consumer: string;
+    executionId: string;
+    id: string;
+    name: string;
+    tags: string[];
+    spaceId: string;
+    parameters: unknown;
+  };
+}
+
+export interface IAlertsAsDataClient<AlertData extends RuleAlertData = never> {
+  create(alert: ReportedAlert & AlertData): void;
+}
+
+export type PublicAlertsClient<
+  AlertData extends RuleAlertData,
+  LegacyState extends AlertInstanceState,
+  LegacyContext extends AlertInstanceContext,
+  ActionGroupIds extends string
+> = Pick<IAlertsAsDataClient<AlertData>, 'create'> & {
+  getAlertLimitValue: () => number;
+  setAlertLimitReached: (reached: boolean) => void;
+  getRecoveredAlerts: () => Array<LegacyAlert<LegacyState, LegacyContext, ActionGroupIds>>;
+};
+
+export class AlertsClient<
+  AlertData extends RuleAlertData,
+  LegacyState extends AlertInstanceState,
+  LegacyContext extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+> implements
+    IAlertsAsDataClient<AlertData>,
+    IAlertsClient<LegacyState, LegacyContext, ActionGroupIds, RecoveryActionGroupId>
+{
+  private rule: AlertRuleSchema | null = null;
+  private alertsFactory: PublicAlertFactory<
+    LegacyState,
+    LegacyContext,
+    WithoutReservedActionGroups<ActionGroupIds, RecoveryActionGroupId>
+  >;
+  private reportedAlerts: Record<string, ReportedAlert & AlertData> = {};
+  private fetchedAlerts: {
+    indices: Record<string, string>;
+    data: Record<string, Alert & AlertData>;
+  };
+
+  constructor(
+    private readonly options: AlertsClientParams<
+      LegacyState,
+      LegacyContext,
+      ActionGroupIds,
+      RecoveryActionGroupId
+    >
+  ) {
+    this.fetchedAlerts = { indices: {}, data: {} };
+    this.alertsFactory = this.options.legacyAlertsClient.getExecutorServices();
+  }
+
+  public async initialize({ rule }: InitializeOpts) {
+    this.setRuleData(rule);
+
+    const context = this.options.ruleType.alerts?.context;
+    const esClient = await this.options.elasticsearchClientPromise;
+
+    const indexTemplateAndPattern = getIndexTemplateAndPattern(context!);
+
+    // Get tracked alert UUIDs to query for
+    const trackedAlerts = this.options.legacyAlertsClient.getTrackedAlerts();
```

Comment:
Queries for the existing alert documents so we can carry over any information if necessary.